File: Algorithm/2023.11/wolfy916/코딩테스트 기출문제/1,2,3 떨어트리기.py
'''
1,2,3 떨어트리기 - 프로그래머스 Lv.4(2023 KAKAO BLINT)
분류 : 트리
'''

def solution(edges, target):
    # [A] 각 회차마다 도착할 리프노드를 탐색하는 함수
    def search(v):
        if len(adjL[v]) == 0: return v
        arrival = search(adjL[v][path[v]])
        path[v] = (path[v] + 1) % len(adjL[v])  # 경로 조정
        return arrival

    # [B] 매번 숫자를 떨어트리며 체킹하는 백트랙킹 함수
    def drop(selected, order):
        nonlocal answer

        # 계산값 모두 일치
        if sum_lst[1:] == target:
            answer = selected
            return True

        # 리프노드 메모이제이션
        if len(orders) > order:
            leaf = orders[order]
        else:
            leaf = search(1)
            orders.append(leaf)

        # 백트랙킹 탐색
        for num in range(1, 4):
            if sum_lst[leaf] + num > target[leaf - 1]: continue
            sum_lst[leaf] += num
            if drop(selected + [num], order + 1): return True
            sum_lst[leaf] -= num

        return False

    # [1] 인접리스트 생성
    adjL = [[] for _ in range(len(target) + 1)]
    edges.sort(key=lambda x: [x[1], x[0]])
    for a, b in edges:
        adjL[a].append(b)  # 단방향 간선 표기

    # [2] 초기 연결된 길 설정
    path = [0] * (len(target) + 1)
    sum_lst = [0] * (len(target) + 1)

    # [3] 숫자 떨어트리며 탐색
    answer = [-1]
    orders = []
    drop([], 0)

    return answer


edges = [[2, 4], [1, 2], [6, 8], [1, 3], [5, 7], [2, 5], [3, 6], [6, 10], [6, 9]]
target = [0, 0, 0, 3, 0, 0, 5, 1, 2, 3]
print(solution(edges, target))

File: Algorithm/2023.11/jonghyo/1,2,3떨어트리기.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges, int[] target) {
        int[] answer = {};
        //자식들을 담아줄 List배열 만들기.
		List <Integer>[]li = new ArrayList[edges.length+1];
		for(int i = 0; i <= edges.length; i++) {
			li[i] = new ArrayList<>();
		}
		//현재 부모가 가르키고있는 부분을 확인해주는 배열하나 생성하기.
		int arr[] = new int[edges.length+1];
		for(int i = 0; i < edges.length; i++) {
			//edges[i][0] = 부모번호 edges[i][1] = 자식번호
			//즉 li의 부모번호에 해당하는 부분에 자식번호 추가해주기.
			li[edges[i][0]].add(edges[i][1]);
		}
		//1부터 자식의 개수 파악하기
		//1개면 0으로 계속 그자리에 있게 하기
		//2개 이상이면 숫자를 주어 변환하게 하기.
		//마지막 리프노드면 -1을 주기.
		for(int i = 1; i < li.length; i++) {
			if(li[i].size() == 0) {
				arr[i] = -2;
			}else if(li[i].size() == 1) {
				arr[i] = -1;
			}else {
				arr[i]  = 0;
			}
			//그러고 마지막에 내부 정렬해주기.
			Collections.sort(li[i]);
		}
		//타겟에 해당하는 수 찾기.
		boolean flag = true;
		int num = 0;
		while(flag) {
			int cnt = 1;
			//현재 넣어야 하는 숫자 찾기.
			while(true) {
				if(li[cnt].size() == 0) {
					//사이즈가 0이면 자식이 없으므로 리프노드 확인 while문 끝내.
					num = cnt;
					break;
				}else if(li[cnt].size() == 1) {
					//자식이 1개 있으므로 cnt를 자식으로 갱신후 다시 while문
					cnt = li[cnt].get(0);
				}else {
					//2개 이상이면 길이 있는 애로 가기.
					int beforecnt = cnt;
					cnt = li[cnt].get(arr[cnt]);
					//다음차례로 변경해주기.
					arr[beforecnt]++;
					if(arr[beforecnt] == li[beforecnt].size())
						arr[beforecnt] = 0;
				}
			}
			//숫자 뭐 넣을지 생각하기.
		}
        return answer;
    }
}